File: controller/health.py
"""
Health checking functionality for HTTP applications.
Performs HTTP health checks and manages container health state.
"""

import aiohttp
import asyncio
import time
import logging
from typing import Dict, List, Optional
from dataclasses import dataclass

logger = logging.getLogger(__name__)

@dataclass
class HealthCheckConfig:
    path: str = "/healthz"
    interval_seconds: int = 5
    timeout_seconds: int = 2
    failure_threshold: int = 3
    success_threshold: int = 1

@dataclass
class HealthStatus:
    is_healthy: bool
    consecutive_failures: int = 0
    consecutive_successes: int = 0
    last_check: float = 0.0
    last_success: float = 0.0
    response_time_ms: float = 0.0

class HealthChecker:
    def __init__(self):
        self.health_configs: Dict[str, HealthCheckConfig] = {}
        self.health_status: Dict[str, HealthStatus] = {}
        self.session: Optional[aiohttp.ClientSession] = None
        self._running = False
        
    async def start(self):
        """Start the health checker background task."""
        if not self._running:
            self.session = aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=10)
            )
            self._running = True
            asyncio.create_task(self._health_check_loop())
            logger.info("Health checker started")
    
    async def stop(self):
        """Stop the health checker and clean up resources."""
        self._running = False
        if self.session:
            await self.session.close()
            self.session = None
        logger.info("Health checker stopped")
    
    def add_target(self, container_id: str, ip: str, port: int, config: HealthCheckConfig = None):
        """Add a container to health monitoring."""
        target_key = f"{ip}:{port}"
        self.health_configs[container_id] = config or HealthCheckConfig()
        self.health_status[container_id] = HealthStatus(is_healthy=False)
        logger.info(f"Added health check target: {target_key}")
    
    def remove_target(self, container_id: str):
        """Remove a container from health monitoring."""
        self.health_configs.pop(container_id, None)
        self.health_status.pop(container_id, None)
        logger.info(f"Removed health check target: {container_id}")
    
    def get_health_status(self, container_id: str) -> Optional[HealthStatus]:
        """Get the current health status of a container."""
        return self.health_status.get(container_id)
    
    def is_healthy(self, container_id: str) -> bool:
        """Check if a container is currently healthy."""
        status = self.health_status.get(container_id)
        return status.is_healthy if status else False
    
    async def _health_check_loop(self):
        """Main health checking loop."""
        while self._running:
            try:
                # Create tasks for all health checks
                tasks = []
                for container_id in list(self.health_configs.keys()):
                    task = asyncio.create_task(self._check_container_health(container_id))
                    tasks.append(task)
                
                # Wait for all health checks to complete
                if tasks:
                    await asyncio.gather(*tasks, return_exceptions=True)
                
                # Wait before next round of checks
                await asyncio.sleep(1)  # Check every second, but respect individual intervals
                
            except Exception as e:
                logger.error(f"Error in health check loop: {e}")
                await asyncio.sleep(5)  # Back off on errors
    
    async def _check_container_health(self, container_id: str):
        """Perform health check for a single container."""
        config = self.health_configs.get(container_id)
        status = self.health_status.get(container_id)
        
        if not config or not status:
            return
        
        # Check if it's time for a health check
        now = time.time()
        if now - status.last_check < config.interval_seconds:
            return
        
        # Get container info (this would be passed from the manager)
        # For now, we'll need to reconstruct this from the container_id
        # In a real implementation, we'd store this mapping
        try:
            # This is a placeholder - in reality we'd get this info from the manager
            container_info = self._get_container_info(container_id)
            if not container_info:
                return
            
            ip, port = container_info["ip"], container_info["port"]
            
            # Perform the health check
            start_time = time.time()
            is_healthy = await self._perform_http_check(ip, port, config)
            response_time = (time.time() - start_time) * 1000  # Convert to ms
            
            # Update status
            status.last_check = now
            status.response_time_ms = response_time
            
            if is_healthy:
                status.consecutive_successes += 1
                status.consecutive_failures = 0
                status.last_success = now
                
                # Mark as healthy if we've had enough consecutive successes
                if status.consecutive_successes >= config.success_threshold:
                    if not status.is_healthy:
                        logger.info(f"Container {container_id} is now healthy")
                    status.is_healthy = True
            else:
                status.consecutive_failures += 1
                status.consecutive_successes = 0
                
                # Mark as unhealthy if we've had too many consecutive failures
                if status.consecutive_failures >= config.failure_threshold:
                    if status.is_healthy:
                        logger.warning(f"Container {container_id} is now unhealthy")
                    status.is_healthy = False
                    
        except Exception as e:
            logger.error(f"Health check failed for container {container_id}: {e}")
            status.consecutive_failures += 1
            status.consecutive_successes = 0
            if status.consecutive_failures >= config.failure_threshold:
                status.is_healthy = False
    
    def _get_container_info(self, container_id: str) -> Optional[Dict]:
        """Get container IP and port info. This would be provided by the manager."""
        # This is a placeholder method. In the real implementation,
        # the manager would provide this information or we'd store it
        # when adding targets.
        return None
    
    async def _perform_http_check(self, ip: str, port: int, config: HealthCheckConfig) -> bool:
        """Perform an HTTP health check against a container."""
        if not self.session:
            return False
        
        try:
            url = f"http://{ip}:{port}{config.path}"
            
            async with self.session.get(
                url,
                timeout=aiohttp.ClientTimeout(total=config.timeout_seconds)
            ) as response:
                # Consider 2xx and 3xx responses as healthy
                return 200 <= response.status < 400
                
        except asyncio.TimeoutError:
            logger.debug(f"Health check timeout for {ip}:{port}")
            return False
        except aiohttp.ClientError as e:
            logger.debug(f"Health check connection error for {ip}:{port}: {e}")
            return False
        except Exception as e:
            logger.warning(f"Unexpected error during health check for {ip}:{port}: {e}")
            return False
    
    def get_all_healthy_containers(self) -> List[str]:
        """Get list of all healthy container IDs."""
        healthy = []
        for container_id, status in self.health_status.items():
            if status.is_healthy:
                healthy.append(container_id)
        return healthy
    
    def get_health_summary(self) -> Dict:
        """Get a summary of all health checks."""
        summary = {
            "total_targets": len(self.health_status),
            "healthy_targets": 0,
            "unhealthy_targets": 0,
            "targets": {}
        }
        
        for container_id, status in self.health_status.items():
            if status.is_healthy:
                summary["healthy_targets"] += 1
            else:
                summary["unhealthy_targets"] += 1
            
            summary["targets"][container_id] = {
                "healthy": status.is_healthy,
                "consecutive_failures": status.consecutive_failures,
                "consecutive_successes": status.consecutive_successes,
                "last_success": status.last_success,
                "response_time_ms": status.response_time_ms
            }
        
        return summary

File: controller_go/health.go
package controller

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"
)

type HealthCheckConfig struct {
	Path             string
	IntervalSeconds  int
	TimeoutSeconds   int
	FailureThreshold int
	SuccessThreshold int
}

type HealthStatus struct {
	IsHealthy           bool
	ConsecutiveFailures int
	ConsecutiveSuccess  int
	LastCheck           time.Time
	LastSuccess         time.Time
	ResponseTimeMS      float64
}

type ContainerInfo struct {
	IP   string
	Port int
}

type HealthChecker struct {
	mu            sync.Mutex
	healthConfigs map[string]HealthCheckConfig
	healthStatus  map[string]*HealthStatus
	containerInfo map[string]ContainerInfo
	running       bool
	callback      func(containerID string, healthy bool)
	wg            sync.WaitGroup
	ctx           context.Context
	cancel        context.CancelFunc
	client        *http.Client
}

// NewHealthChecker creates a new health checker
func NewHealthChecker() *HealthChecker {
	ctx, cancel := context.WithCancel(context.Background())
	return &HealthChecker{
		healthConfigs: make(map[string]HealthCheckConfig),
		healthStatus:  make(map[string]*HealthStatus),
		containerInfo: make(map[string]ContainerInfo),
		ctx:           ctx,
		cancel:        cancel,
		client: &http.Client{
			Timeout: 10 * time.Second,
		},
	}
}

// SetHealthChangeCallback sets a callback function when a container health status changes
func (hc *HealthChecker) SetHealthChangeCallback(cb func(string, bool)) {
	hc.callback = cb
}

// AddTarget adds a new container for health monitoring
func (hc *HealthChecker) AddTarget(containerID, ip string, port int, cfg HealthCheckConfig) {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	if cfg.Path == "" {
		cfg.Path = "/healthz"
	}
	if cfg.IntervalSeconds == 0 {
		cfg.IntervalSeconds = 5
	}
	if cfg.TimeoutSeconds == 0 {
		cfg.TimeoutSeconds = 2
	}
	if cfg.FailureThreshold == 0 {
		cfg.FailureThreshold = 3
	}
	if cfg.SuccessThreshold == 0 {
		cfg.SuccessThreshold = 1
	}

	hc.healthConfigs[containerID] = cfg
	hc.healthStatus[containerID] = &HealthStatus{IsHealthy: false}
	hc.containerInfo[containerID] = ContainerInfo{IP: ip, Port: port}

	log.Printf("Added health check target: %s:%d for container %s", ip, port, containerID)
}

// RemoveTarget removes a container from health monitoring
func (hc *HealthChecker) RemoveTarget(containerID string) {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	delete(hc.healthConfigs, containerID)
	delete(hc.healthStatus, containerID)
	delete(hc.containerInfo, containerID)
	log.Printf("Removed health check target: %s", containerID)
}

// IsHealthy returns whether a container is healthy
func (hc *HealthChecker) IsHealthy(containerID string) bool {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	if status, ok := hc.healthStatus[containerID]; ok {
		return status.IsHealthy
	}
	return false
}

// Start begins the health checking loop in background goroutines
func (hc *HealthChecker) Start() {
	if hc.running {
		return
	}
	hc.running = true
	hc.wg.Add(1)
	go hc.loop()
	log.Println("Health checker started")
}

// Stop stops the health checking
func (hc *HealthChecker) Stop() {
	hc.cancel()
	hc.wg.Wait()
	hc.running = false
	log.Println("Health checker stopped")
}

// loop runs continuous health checks
func (hc *HealthChecker) loop() {
	defer hc.wg.Done()

	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-hc.ctx.Done():
			return
		case <-ticker.C:
			hc.runChecks()
		}
	}
}

// runChecks launches concurrent health checks for all containers
func (hc *HealthChecker) runChecks() {
	hc.mu.Lock()
	ids := make([]string, 0, len(hc.healthConfigs))
	for id := range hc.healthConfigs {
		ids = append(ids, id)
	}
	hc.mu.Unlock()

	var wg sync.WaitGroup
	for _, id := range ids {
		wg.Add(1)
		go func(cid string) {
			defer wg.Done()
			hc.checkContainer(cid)
		}(id)
	}
	wg.Wait()
}

// checkContainer performs health check for a single container
func (hc *HealthChecker) checkContainer(containerID string) {
	hc.mu.Lock()
	cfg, cfgOk := hc.healthConfigs[containerID]
	status, stOk := hc.healthStatus[containerID]
	info, infoOk := hc.containerInfo[containerID]
	hc.mu.Unlock()

	if !cfgOk || !stOk || !infoOk {
		return
	}

	now := time.Now()
	if now.Sub(status.LastCheck) < time.Duration(cfg.IntervalSeconds)*time.Second {
		return
	}

	status.LastCheck = now

	start := time.Now()
	healthy := hc.performHTTPCheck(info, cfg)
	elapsed := time.Since(start).Milliseconds()
	status.ResponseTimeMS = float64(elapsed)

	if healthy {
		status.ConsecutiveSuccess++
		status.ConsecutiveFailures = 0
		status.LastSuccess = now
		if !status.IsHealthy && status.ConsecutiveSuccess >= cfg.SuccessThreshold {
			status.IsHealthy = true
			log.Printf("Container %s is now healthy", containerID)
			if hc.callback != nil {
				hc.callback(containerID, true)
			}
		}
	} else {
		status.ConsecutiveFailures++
		status.ConsecutiveSuccess = 0
		if status.IsHealthy && status.ConsecutiveFailures >= cfg.FailureThreshold {
			status.IsHealthy = false
			log.Printf("Container %s is now unhealthy", containerID)
			if hc.callback != nil {
				hc.callback(containerID, false)
			}
		}
	}
}

// performHTTPCheck performs an HTTP GET and returns true if response is 2xx–3xx
func (hc *HealthChecker) performHTTPCheck(info ContainerInfo, cfg HealthCheckConfig) bool {
	url := fmt.Sprintf("http://%s:%d%s", info.IP, info.Port, cfg.Path)
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(cfg.TimeoutSeconds)*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		log.Printf("Error creating request: %v", err)
		return false
	}

	resp, err := hc.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode >= 200 && resp.StatusCode < 400
}

// GetHealthStatus returns the current health status of a container
func (hc *HealthChecker) GetHealthStatus(containerID string) *HealthStatus {
	hc.mu.Lock()
	defer hc.mu.Unlock()
	return hc.healthStatus[containerID]
}

// GetAllHealthyContainers returns all healthy container IDs
func (hc *HealthChecker) GetAllHealthyContainers() []string {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	var healthy []string
	for id, status := range hc.healthStatus {
		if status.IsHealthy {
			healthy = append(healthy, id)
		}
	}
	return healthy
}

// GetHealthSummary returns a summary of all health checks
func (hc *HealthChecker) GetHealthSummary() map[string]interface{} {
	hc.mu.Lock()
	defer hc.mu.Unlock()

	summary := map[string]interface{}{
		"total_targets":     len(hc.healthStatus),
		"healthy_targets":   0,
		"unhealthy_targets": 0,
		"targets":           map[string]interface{}{},
	}

	for id, s := range hc.healthStatus {
		target := map[string]interface{}{
			"healthy":              s.IsHealthy,
			"consecutive_failures": s.ConsecutiveFailures,
			"consecutive_success":  s.ConsecutiveSuccess,
			"last_success":         s.LastSuccess,
			"response_time_ms":     s.ResponseTimeMS,
		}
		if s.IsHealthy {
			summary["healthy_targets"] = summary["healthy_targets"].(int) + 1
		} else {
			summary["unhealthy_targets"] = summary["unhealthy_targets"].(int) + 1
		}
		summary["targets"].(map[string]interface{})[id] = target
	}
	return summary
}
